index.js

```diff
@@ -49,51 +52,31 @@ exports.handler = async (event) => {
       const total = bytesDownloaded.reduce((a, b) => a + b, 0)
       const totalSeconds = arr.bytesToSeconds(total)
       const totalPercent = arr.bytesToPercent(total)
-      const hasSeconds = totalSeconds >= minSeconds
-      const hasPercent = totalPercent >= minPercent
-      const overallReason = hasSeconds ? 'seconds' : (hasPercent ? 'percent' : false)
-
-      // start waiting for impressions
-      let waiters = []
-      if (overallReason) {
-        const imp = {...bytesData, bytes: total, seconds: totalSeconds, percent: totalPercent}
-        waiters.push(kinesis.putImpressionLock(redis, imp))
-      } else {
-        waiters.push(false)
+      if (totalSeconds >= minSeconds || totalPercent >= minPercent) {
+        kinesisRecords.push({...bytesData, bytes: total, seconds: totalSeconds, percent: totalPercent})
       }
 
       // check which segments have been FULLY downloaded
-      waiters = waiters.concat(arr.segments.map(async ([firstByte, lastByte], idx) => {
+      arr.segments.forEach(([firstByte, lastByte], idx) => {
         if (arr.isLoggable(idx) && range.complete(firstByte, lastByte)) {
-          const imp = {...bytesData, segment: idx}
-          return await kinesis.putImpressionLock(redis, imp)
-        } else {
-          return false
+          kinesisRecords.push({...bytesData, segment: idx})
         }
-      }))
+      })
+    }))
 
-      const [overallImpression, ...segmentImpressions] = await Promise.all(waiters)
-      return {
-        segments: segmentImpressions,
-        overall: overallImpression ? overallReason : false,
-        overallBytes: total,
-      }
-    })
+    // put kinesis records in batch
+    const results = await kinesis.putWithLock(redis, kinesisRecords)
```

Comment:
here they all get put with lock to kinesis, and this must be where the 10s timeout is used